backend/app/notify/email.go

```diff
@@ -0,0 +1,162 @@
+		keepAlive: params.KeepAlive,
+		sendChan:  make(chan *gomail.Message),
+		errChan:   make(chan error),
+	}
+
+	if res.keepAlive == 0 {
+		res.keepAlive = emailConnectionKeepAlive
+	}
+
+	log.Printf(
+		"[DEBUG] create new email notifier for server %s with user %s, keepalive=%s",
+		res.server, res.username, res.keepAlive)
+
+	// Test connection before starting a daemon.
+	tmpConn, err := gomail.NewDialer(res.server, res.port, res.username, res.password).Dial()
+	if err != nil {
+		return &res, errors.Errorf(
+			"[WARN] error establishing test connecting to '%s':%d with username '%s': %s",
+			res.server, res.port, res.username, err)
+	}
+	if err = tmpConn.Close(); err != nil {
+		return &res, errors.Errorf(
+			"[WARN] error closing test connection to %s:%d: %s",
+			res.server, res.port, err)
+	}
+
+	// Activate server goroutine.
+	go res.activate()
+	// Close the channel to stop the mail daemon
+	go func() { <-ctx.Done(); close(res.sendChan) }()
+
+	return &res, nil
+}
+
+// Send email from request to address in settings
+func (e *Email) Send(ctx context.Context, req request) error {
+	log.Printf("[DEBUG] send notification via %s, comment id %s", e, req.comment.ID)
+
+	messageBody := prepareBody(req)
+
+	// Create message.
+	m := gomail.NewMessage()
+	// TODO: figure out where "from" and "to" addresses come from
+	//m.SetHeader("From", "no-reply@example.com")
+	//m.SetAddressHeader("To", req.Address, req.Name)
+	m.SetHeader("Subject", fmt.Sprintf("New comment for \"%s\"", req.comment.PostTitle))
+	m.SetBody("text/html", messageBody)
+
+	// Wait for ability to send message and return error from error channel after sending it.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case e.sendChan <- m:
+		return <-e.errChan
+	}
+}
+
+func (e *Email) String() string {
+	return fmt.Sprintf("email: '%s'@'%s':%d", e.username, e.server, e.port)
+}
+
+//prepareBody generates email message text based on request
+func prepareBody(req request) string {
+	from := req.comment.User.Name
+	if req.comment.ParentID != "" {
+		from += " → " + req.parent.User.Name
+	}
+	link := fmt.Sprintf("↦ [original comment](%s)", req.comment.Locator.URL+uiNav+req.comment.ID)
+	if req.comment.PostTitle != "" {
+		link = fmt.Sprintf("↦ [%s](%s)", req.comment.PostTitle, req.comment.Locator.URL+uiNav+req.comment.ID)
+	}
+	body := fmt.Sprintf("%s\n\n%s\n\n%s", from, req.comment.Orig, link)
+	return body
+}
+
+func (e *Email) activate() {
+	d := gomail.NewDialer(e.server, e.port, e.username, e.password)
+
+	var s gomail.SendCloser
+	var err error
+	open := false
+	for {
+		select {
+		case m, ok := <-e.sendChan:
+			if !ok {
+				close(e.errChan)
+				return
+			}
+			if !open {
+				if s, err = d.Dial(); err != nil {
+					// Problems with connection, returning error and considering connection not established.
+					e.errChan <- errors.Errorf(
+						"error connecting to %s:%d with username %s: %s",
+						e.server, e.port, e.username, err)
+					break
```

Comment:
breaks inside of for look within select (or switch) are very confusing. It takes time to figure what it actually breaks. Will be better to avoid it

Reply:
It's basically a reply: stop whatever we do and shut down the server because of the error. I'll do it differently after connecting to email server would be operated with retries.